src/main/kotlin/org/rust/ide/refactoring/generateConstructor/GenerateConstructorHandler.kt

```diff
@@ -31,37 +31,59 @@ class GenerateConstructorAction : CodeInsightAction() {
         generateConstructorHandler.isValidFor(editor, file)
 }
 
+data class Context(val struct: RsStructItem, val implBlock: RsImplItem? = null)
+
 class GenerateConstructorHandler : LanguageCodeInsightActionHandler {
-    override fun isValidFor(editor: Editor, file: PsiFile): Boolean = getStructItem(editor, file) != null
+    override fun isValidFor(editor: Editor, file: PsiFile): Boolean = getContext(editor, file) != null
 
     override fun startInWriteAction() = false
-    private fun getStructItem(editor: Editor, file: PsiFile): RsStructItem? =
-        file.findElementAt(editor.caretModel.offset)?.ancestorOrSelf()
+    private fun getContext(editor: Editor, file: PsiFile): Context? {
+        val element = file.findElementAt(editor.caretModel.offset)?: return null
+        val struct = element.ancestorOrSelf<RsStructItem>()
+        return if (struct != null) {
+            Context(struct)
+        }
+        else {
+            val impl = element.ancestorOrSelf<RsImplItem>() ?: return null
+            if (!isImplBlockValid(impl)) return null
+
+            Context(impl.typeReference?.baseType?.path?.reference?.resolve() as? RsStructItem ?: return null, impl)
+        }
+    }
 
     override fun invoke(project: Project, editor: Editor, file: PsiFile) {
-        val structItem = getStructItem(editor, file) ?: return
-        generateConstructorBody(structItem, editor)
+        val context = getContext(editor, file) ?: return
+        generateConstructorBody(context, editor)
     }
 
-    private fun generateConstructorBody(structItem: RsStructItem, editor: Editor) {
+    private fun generateConstructorBody(context: Context, editor: Editor) {
         checkWriteAccessNotAllowed()
-        val chosenFields = showConstructorArgumentsChooser(structItem.project, structItem) ?: return
+        val chosenFields = showConstructorArgumentsChooser(context.struct.project, context.struct) ?: return
         runWriteAction {
-            insertNewConstructor(structItem, chosenFields, editor)
+            insertNewConstructor(context.struct, context.implBlock, chosenFields, editor)
         }
     }
 
-    private fun insertNewConstructor(structItem: RsStructItem, selectedFields: List<ConstructorArgument>, editor: Editor) {
+    private fun isImplBlockValid(impl: RsImplItem): Boolean {
+        if (impl.traitRef != null) return false
+
+        return impl.members?.functionList?.find { it.name == "new" } == null
```

Comment:
It doesn't work because you can have several `impl` blocks for the same item.
You need to check that you don't have `new` method for struct in whole project.

@vlad20012 do we have some API to check method existence for some type without taking into account type params? Or we should use `org.rust.lang.core.resolve.indexes.RsImplIndex.Companion#findPotentialImpls` directly?
Note, it may be quite heavy operation so you need to move these check from `GenerateConstructorHandler#isValidFor` to `GenerateConstructorHandler#invoke` and abort action if needed

Reply:
You're right, but before this PR the intention also didn't check if another constructor exists. So it's an added (best-effort) check that shouldn't introduce regressions. Therefore I think that we could keep it even if we don't use a global lookup of impls of some kind. A global lookup would be of course nicer :-)